### PR TITLE
Add settings for creating and duplicating spaces

### PR DESCRIPTION
**Pull request by martinbrose**
References https://github.com/huggingface/huggingface_hub/issues/1457

**Key Features**

- Modifies hf_api.py to add space-related settings to create_repo.
- Modifies hf_api.py to add additional settings to duplicate_space.

### CODE REVIEW

**Comment by martinbrose** [comment]
I tried to come up with tests for duplicating spaces.
But tbh, I don't think I have the best approach. I'm getting account rate-limited.
So I believe I'm a little bit too resource-hungry with the test case.

I'll probably leave this with you @Wauplin if you don't mind. You probably have the best idea of how to mock this...

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -5357,6 +5391,21 @@ def duplicate_space(
...
+            sleepTimeSeconds (`int`, *optional*):
+                Number of seconds of inactivity to wait before a Space is put to sleep. Set to `-1` if you don't want
+                your Space to sleep (default behavior for upgraded hardware). For free hardware, you can't configure
+                the sleep time (value is fixed to 48 hours of inactivity).
+                See https://huggingface.co/docs/hub/spaces-gpus#sleep-time for more details.
+            secrets (`List[Dict[str, str]]`, *optional*):
+                Secrets allow to set secret keys or tokens to a Space without hardcoding them.
+                For more details, see https://huggingface.co/docs/hub/spaces-overview#managing-secrets.
+            variables (`List[Dict[str, str]]`, *optional*):
+                Variables allow to set environment variables to a Space without hardcoding them.
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:5407]:
```suggestion
                A list of public environment variables to set in your Space. Each item is in the form `{"key": ..., "value": ..., "description": ...}` where description is optional.
```

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -5357,6 +5391,21 @@ def duplicate_space(
...
+                Choice of Hardware. Example: `"t4-medium"`. See [`SpaceHardware`] for a complete list.
+            storage (`SpaceStorage` or `str`, *optional*):
+                Choice of persistent storage tier. Example: `"small"`. See [`SpaceStorage`] for a complete list.
+            sleepTimeSeconds (`int`, *optional*):
+                Number of seconds of inactivity to wait before a Space is put to sleep. Set to `-1` if you don't want
+                your Space to sleep (default behavior for upgraded hardware). For free hardware, you can't configure
+                the sleep time (value is fixed to 48 hours of inactivity).
+                See https://huggingface.co/docs/hub/spaces-gpus#sleep-time for more details.
+            secrets (`List[Dict[str, str]]`, *optional*):
+                Secrets allow to set secret keys or tokens to a Space without hardcoding them.
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:5404]:
```suggestion
                A list of secret keys to set in your Space. Each item is in the form `{"key": ..., "value": ..., "description": ...}` where description is optional.
```

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -5357,6 +5391,21 @@ def duplicate_space(
                 Hugging Face token. Will default to the locally saved token if not provided.
             exist_ok (`bool`, *optional*, defaults to `False`):
                 If `True`, do not raise an error if repo already exists.
+            hardware (`SpaceHardware` or `str`, *optional*):
+                Choice of Hardware. Example: `"t4-medium"`. See [`SpaceHardware`] for a complete list.
+            storage (`SpaceStorage` or `str`, *optional*):
+                Choice of persistent storage tier. Example: `"small"`. See [`SpaceStorage`] for a complete list.
+            sleepTimeSeconds (`int`, *optional*):
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:5398]:
```suggestion
            sleep_time (`int`, *optional*):
```

Same here, `sleep_time` would be best :)

**Review by Wauplin — COMMENTED** [review]
That's a very good quality PR, thanks for the work @martinbrose ! I have added comments about renaming/slightly rephrasing but that's it. About the `duplicate_space` tests, I think you can use the exact same approach as the mocked `create_repo` tests. With mock tests you cannot be rate-limited since http requests are not made to the Hub. I'm fine with adding them myself if you prefer. 

Also to complete this PR, would you mind updating the [Manage your Space](https://huggingface.co/docs/huggingface_hub/guides/manage-spaces#a-simple-example-configure-secrets-and-hardware) guide in the docs? You can update them in the [`docs/` folder](https://github.com/huggingface/huggingface_hub/blob/main/docs/source/en/guides/manage-spaces.md). There is a section in a guide _"Bonus: request hardware when creating the Space!"_. I would update it to show how to set hardware + sleep time + secrets + variables at the same time at repo creation. There is also a guide for [duplicating Spaces](https://huggingface.co/docs/huggingface_hub/v0.16.3/en/guides/repository#clone-a-repository-only-for-spaces) that would benefit from a mention of the new arguments you've added. I would keep the existing minimal example and a another complete example just below and explain in a few words why it's useful. Let me know what you think.

Thanks in advance!

File: src/huggingface_hub/hf_api.py [hunk 4]
```diff
@@ -2365,7 +2365,11 @@ def create_repo(
         repo_type: Optional[str] = None,
         exist_ok: bool = False,
         space_sdk: Optional[str] = None,
-        space_hardware: Optional[str] = None,
+        space_hardware: Optional[SpaceHardware] = None,
+        space_storage: Optional[SpaceStorage] = None,
+        space_sleepTimeSeconds: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (martinbrose) [reply src/huggingface_hub/hf_api.py]: Apologies, a slight oversight on my side. Corrected.

File: src/huggingface_hub/hf_api.py (hunk 3 above)
Comment (martinbrose) [reply src/huggingface_hub/hf_api.py]: Changed this to `sleep_time` throughout. Also updated the test case.

**Comment by martinbrose** [comment]
Thanks for the review and the kind recognition.

I addressed all the code review points and added the documentation.
I'll work on the duplication test cases in the next few days.

**Review by Wauplin — COMMENTED** [review]
Nice improvement in the docs! Waiting for the final tests for duplicate_space before merging :)

**Comment by martinbrose** [comment]
Hi @Wauplin, I think I was overthinking the test case... we are only testing if the API call is correct.

This is now addressed in the additional test case.

**Review by Wauplin — APPROVED** [review]
Great job @martinbrose! Thanks for the last test, it's perfect like this. The PR is fine to merge now :)

**Pull request merged**